lib/auth/webauthncli/fido2.go

```diff
@@ -562,54 +582,277 @@ func runOnFIDO2Devices(
 	filter deviceFilterFunc,
 	deviceCallback deviceCallbackFunc,
 ) error {
-	// About to select, prompt user.
+	locs, err := fidoDeviceLocations()
+	if err != nil {
+		return trace.Wrap(err, "device locations")
+	}
+	if len(locs) == 0 {
+		return trace.Wrap(errors.New("no security keys found"))
+	}
+
+	devices, devicesC, err := startDevices(locs, filter, deviceCallback, prompt)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var receiveCount int
+	defer func() {
+		// Cancel all in-flight requests, if any.
+		devices.cancelAll(nil /* except */)
+
+		// Give the devices some time to tidy up, but don't wait forever.
+		maxWait := time.NewTimer(fido2DeviceMaxWait)
+		defer maxWait.Stop()
+
+		for receiveCount < devices.len() {
+			select {
+			case <-devicesC:
+				receiveCount++
+			case <-maxWait.C:
+				log.Debugf("FIDO2: Abandoning device goroutines after %s", fido2DeviceMaxWait)
+				return
+			}
+		}
+		log.Debug("FIDO2: Device goroutines exited cleanly")
+	}()
+
+	// First "interactive" response wins.
+	for receiveCount < devices.len() {
+		select {
+		case err := <-devicesC:
+			receiveCount++
+
+			// Keep going on cancels or non-interactive errors.
+			if errors.Is(err, libfido2.ErrKeepaliveCancel) || errors.Is(err, &nonInteractiveError{}) {
+				log.Debugf("FIDO2: Got cancel or non-interactive device error: %v", err)
+				continue
+			}
+
+			return trace.Wrap(err)
+
+		case <-ctx.Done():
+			return trace.Wrap(ctx.Err())
+		}
+	}
+	return trace.Wrap(errors.New("all MFA devices failed"))
+}
+
+func startDevices(
+	locs []*libfido2.DeviceLocation,
+	filter deviceFilterFunc,
+	deviceCallback deviceCallbackFunc,
+	prompt runPrompt) (devices *openedDevices, devicesC <-chan error, err error) {
+	fidoDevs := make([]FIDODevice, 0, len(locs))
+	openDevs := make([]*openedDevice, 0, len(locs))
```

Comment:
```suggestion
	prompt runPrompt) (devices *openedDevices, devicesC <-chan error, err error) {
	
	fidoDevs := make([]FIDODevice, 0, len(locs))
	openDevs := make([]*openedDevice, 0, len(locs))
```
nit: make it visually easier to see where the function body begins since this is a multiline func signature

Reply:
TBH I usually kill spaces like these. I added a comma to the last parameter, which causes a difference in indentation, so hopefully that's better:

```diff
-	prompt runPrompt) (devices *openedDevices, devicesC <-chan error, err error) {
+	prompt runPrompt,
+) (devices *openedDevices, devicesC <-chan error, err error) {
	fidoDevs := make([]FIDODevice, 0, len(locs))
	openDevs := make([]*openedDevice, 0, len(locs))
```